### DISCUSSION TITLE
Architecture issue on Unicode and PC Screen Font

### DISCUSSION THREAD

**Discussion by valentinbreiz** [CosmosOS/Cosmos#1601] [Q&A]
Hello everyone I have an architecture design issue
I'm trying to implement unicode with a custom PC Screen Font. It works well but I don't know if I make it the right way. The problem is that the font doesn't support all unicode characters (that is normal unicode is huge), so some characters has differents offsets in the font data and unicode value.

For exemple for a space (' '), unicode is _U+0020_ and my font offset is _0x20_, they are the sames no problem. But for example if I want to use the double bar ('║') which is _U+2551_ for unicode and at _0xD1_ address in my font data: they do not patch.

My technique seems for me a bit tricky: I just search for unicode correspondance in the unicode map (parsed thought the PCScreenFont class) like above:
``` csharp
public void WriteByte(char ch)
{
    DrawFilledRectangle(backpen, Kernel.AConsole.X * font.Width + 1, Kernel.AConsole.Y * font.Height, font.Width - 1, font.Height);

    if (ch > 0x7E) //max font. After 0x7E font data and unicode values are not the same
    {
        foreach (var map in font._UnicodeMappings)
        {
            foreach (var unicodevalue in map.UnicodeCharacters)
            {
                if (unicodevalue == ch)
                {
                    canvas.DrawChar((char)map.FontPosition, font, frontpen, Kernel.AConsole.X * font.Width, Kernel.AConsole.Y * font.Height);
                    //correspondance.FontPosition is 0x20 for the space example
                }
            }
        }
    }
    else
    {
        canvas.DrawChar(ch, font, frontpen, Kernel.AConsole.X * font.Width, Kernel.AConsole.Y * font.Height);
    }
}
```
Do you think is it right like this or should I make an array of chars with the max unicode lenght supported by the font and place the font data in each unicode offset?

**Comment by ghost** [comment]
It may be that it's unrelated to the architecture- many unicode characters are not supported at all. It could also be the font itself.
